### PR TITLE
http: moving strict header checking into the http/1.1 codec

### PR DESCRIPTION
**Pull request by alyssawilk**
As discovered on #7306 runtime checks don't work well outside of worker threads.
This leverages the fixed over in #7695 to move strict header checking to the codec.

Risk Level: Medium (messing with runtime, HTTP/1 parsing)
Testing: previously failing integration tests now pass
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

**Comment by alyssawilk** [comment]
@mattklein123 thoughts on the options above?

File: include/envoy/thread_local/thread_local.h [hunk 1]
```diff
@@ -28,6 +28,14 @@ class Slot {
 public:
   virtual ~Slot() = default;
 
+  /**
+   * This is a helper, mainly used for test code, to determine if a given thread
+   * has a valid slot (i.e. if it is a worker thread) or not.
+   *
+   * @return true if this is a valid slot (called from a worker thread).
+   */
+  virtual bool valid() PURE;
```
Comment (mattklein123) [inline include/envoy/thread_local/thread_local.h]: nit: `currentThreadInitialized()` or similar?

**Comment by alyssawilk** [comment]
I think we can actually do one better, and make snapshot access actually const - looks like nothing but test code uses non-const and I can fix that up.  If so we can combine 2 and 3 for a fairly simple solution.
I'll go ahead and land that as a standone PR now, to see if it causes anyone trouble, then go fix this one up to use it.

**Comment by alyssawilk** [comment]
Ok, I think this works better, but I want to see how CI is and probably do one more clean up pass before I remove the waiting tag.
Matt one question if you have a few, do you think this is large enough I should split the runtime change out from the HTTP feature port or leave them together?

**Comment by mattklein123** [comment]
> Matt one question if you have a few, do you think this is large enough I should split the runtime change out from the HTTP feature port or leave them together?

Optimally I think it's worth splitting it out if you don't mind, but it's not a big deal to me either way.

**Comment by alyssawilk** [comment]
I prefer smaller safer PRs so I'll close this off for a few and reopen once the runtime-only change has landed.

**Pull request merged**